### ISSUE TITLE
Controls: boolean control throws a red warning

### ISSUE THREAD

**Issue by matheo** [storybookjs/storybook#10967]
**Describe the bug**

Playing with `"@storybook/addon-controls": "6.0.0-beta.17"` I got this warning after clicking a boolean control, which is the `inverseColors` input (at the beggining of the screenshot):

![image](https://user-images.githubusercontent.com/260185/83174657-4f31be80-a0e0-11ea-9839-239a7faab0af.png)

Upcoming clicks in the checkbox doesn't throw this,
as it seems that the checkbox is already changed to controlled.

**System:**

```
Environment Info:

  System:
    OS: Linux 5.4 Ubuntu 20.04 LTS (Focal Fossa)
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  Binaries:
    Node: 10.18.0 - ~/.nvm/versions/node/v10.18.0/bin/node
    Yarn: 1.22.4 - /usr/bin/yarn
    npm: 6.14.2 - ~/.nvm/versions/node/v10.18.0/bin/npm
  Browsers:
    Chrome: 83.0.4103.61
    Firefox: 76.0.1
  npmPackages:
    @storybook/addon-a11y: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/addon-actions: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/addon-controls: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/addon-docs: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/addon-links: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/addon-viewport: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/addons: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/angular: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/cli: 6.0.0-beta.17 => 6.0.0-beta.17 
    @storybook/source-loader: 6.0.0-beta.17 => 6.0.0-beta.17
```

**Comment by matheo** [comment]
The same happened with a `number` control:

![image](https://user-images.githubusercontent.com/260185/83209921-1bc25480-a11f-11ea-80b0-82b89425e397.png)

**Comment by shilman** [comment]
Do you have a public repro for this, or can you add one to `official-storybook`? I played around with this for a minute and couldn't repro in https://raw.githubusercontent.com/storybookjs/storybook/next/examples/official-storybook/stories/addon-docs/props.stories.mdx

**Comment by matheo** [comment]
@shilman wooops,
I forked the `ControlsPanel` with the exact same code and registered a panel just like `@storybook/addon-controls`. I thought it should behave the same as the official plugin, but it doesn't then :(

**Comment by shilman** [comment]
@matheo hmmm did you try recently? might have gotten fixed since you filed this

**Comment by matheo** [comment]
yeah, I just compared the code of `addon-controls` and my fork, and they are basically the same.

I will dig a little bit in the argTypes of the controls that are failing to see if there's something missing there.

**Comment by shilman** [comment]
Thanks! It wouldn't surprise me if there was still a bug, just having trouble repro-ing it! 😄

**Comment by shilman** [comment]
@matheo I found a repro case in `examples/angular-cli`, so I don't need one anymore. Fix coming soon!

**Comment by matheo** [comment]
I checked some stories of `angular-cli` and there was no error thrown, so I got surprised :P

**Comment by shilman** [comment]
ZOMG!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.0.21 containing PR #12322 that references this issue. Upgrade today to try it out!

Closing this issue. Please re-open if you think there's still more to do.

**Comment by gregfagan** [comment]
I updated to v6.0.21 and this is still happening for me.

**Comment by shilman** [comment]
@gregfagan Do you have a repro repo you can share?

**Comment by gregfagan** [comment]
> @gregfagan Do you have a repro repo you can share?

Sorry, not at the moment; this is in a private repo

**Comment by gregfagan** [comment]
Actually this might have been a stale cache or some other problem. It is no longer throwing errors for my. My mistake.

**Comment by monolithed** [comment]
The same problem with 6.1.0-rc.0.

```
...
        rotate: {
            description: 'Повернуть элемент (не работает в IE 9)',
            table: {
                type: {
                    summary: 'number'
                },
                defaultValue: {
                    summary: 0
                }
            },
            control: {
                type: 'number'
            }
        },
...
```

<img width="466" alt="Снимок экрана 2020-11-13 в 0 23 23" src="https://user-images.githubusercontent.com/803674/98998009-81c45c00-2546-11eb-860c-f9fb8e3e3e06.png">

**Comment by shilman** [comment]
@monolithed does it fix it to add: 

```
        rotate: {
            description: 'Повернуть элемент (не работает в IE 9)',
            defaultValue: 0,
            table: { ... },
            control: { ... },
        }
```

**Comment by monolithed** [comment]
I have already tried it before. it can be reproduced with the simplest component:

```tsx
const Input = ({...props}) => (<Input {...props} />);

...
        value: {
            description: 'Input',
            defaultValue: 0,
            table: {
                type: {
                    summary: 'number'
                },
            },
            control: {
                type: 'number'
            }
        }
...

```

**Comment by fast-reflexes** [comment]
Just take the efault stories inclue in the storybook release, a property `testThis?: number` to it, then assign to `testThis` in only one of the test cases an then fire up storybook and toggle between the buttons to see this behaviour. I have v. 6.0.28.

**Comment by kevinzang** [comment]
I get this error on 6.1.18, but manually setting `defaultValue: 0` _does_ fix it. The autogenerated `argTypes` from react-docgen has `defaultValue: '0'`.

**Comment by shilman** [comment]
@kevinzang can you try in 6.2-beta? we recently fixed a bunch of bugs around default value handling

**Comment by kevinzang** [comment]
@shilman with 6.2.0-beta.3 I can still reproduce this error in two ways:

1. With a **required** number prop, I see the error as soon as I navigate from another story (Default) to one that uses controls (WithControls).

<details><summary>Sample code</summary>

foo.js
```
import PropTypes from 'prop-types';
import React from 'react';

export default class Foo extends React.Component {
    static propTypes = {
        foo: PropTypes.number.isRequired,
    };

    render() {
        return this.props.foo;
    }
}
```

foo.stories.js
```
import React from 'react';

import Foo from './foo';

export default {title: 'Foo', component: Foo};

export const Default = () => <Foo foo={10} />;

export const WithControls = args => <Foo {...args} />;

WithControls.args = {foo: 10};

```

</details>

2. With an **optional** number prop, the above issue doesn't happen (which is probably what you fixed), but the error still happens when I empty the control.

<details><summary>Sample code</summary>

Change the Foo component from the above sample code to

```
static propTypes = {
    foo: PropTypes.number,
};

static defaultProps = {
    foo: 0,
};
```

</details>

<details><summary>Demo</summary>

![controlled-to-uncontrolled](https://user-images.githubusercontent.com/6496295/109235249-2c712b00-7782-11eb-8f75-776326b67342.gif)

</details>

**Comment by anilanar** [comment]
I can reproduce this using v6.2.5 with `react-docgen-typescript` via 0 config setup, for React properties that are optional and whose types are `number`.

**Comment by hieucd04** [comment]
I can reproduce this too using v6.2.9 with `react-docgen-typescript`
I used the below piece of code:
```
// file name: component.stories.mdx

<Story
    name="Bug"
    argTypes={{
        test: {
            control: {
                type: "number"
            }
        }
    }}
>
    {args => <div>bug</div>}
</Story>
```
Screenshot:

![Capture](https://user-images.githubusercontent.com/26107905/125505048-dfeb56c3-b957-4163-bddd-1b6cec97b69e.PNG)

Initially, Storybook rendered a blank numeric input field control and the console was empty. However, when I typed `1` in the numeric input field control the error appeared.

**Comment by simlmx** [comment]
Adding a default value for the option fixes it for me:
```javascript
 export default {
   component: SomeComponent,
   title: "Components/SomeComponent",
   argTypes: {
     myNumber: {
       control: {
         type: "range",
         min: 2,
         max: 7,
       },
     }
   },
   // This next lines is the fix!
   args: {
     myNumber: 5,
   },
 };
```

**Comment by valentinpalkovic** [comment]
Closing as outdated. This was filed against `6.0.0-beta.17` and references the standalone `@storybook/addon-controls` package, which was merged into core in Storybook 9.0. The Controls implementation has been rewritten since; if a similar React "uncontrolled-to-controlled" warning still appears in v10, please open a new issue with a current repro.
